firebase-common/src/main/java/com/google/firebase/Timestamp.kt

```diff
@@ -0,0 +1,143 @@
+ *
+ * Represented as seconds and fractions of seconds at nanosecond resolution in UTC Epoch time. It is
+ * encoded using the Proleptic Gregorian Calendar which extends the Gregorian calendar backwards to
+ * year one. Furthermore,It is encoded assuming all minutes are 60 seconds long, specifically leap
+ * seconds are "smeared" so that no leap second table is needed for interpretation. Range is from
+ * 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z. By restricting to that range, we ensure
+ * that we can convert to and from RFC 3339 date strings.
+ *
+ * @see <a href="https://git.page.link/timestamp-proto">Timestamp</a>The ref timestamp definition
+ */
+class Timestamp : Comparable<Timestamp>, Parcelable {
+  val seconds: Long
+  val nanoseconds: Int
+
+  /**
+   * Creates a new [Timestamp].
+   *
+   * @param seconds represents seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z. Must be
+   * from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z inclusive.
+   * @param nanoseconds represents non-negative fractions of a second at nanosecond resolution.
+   * Negative second values with fractions must still have non-negative nanoseconds values that
+   * count forward in time. Must be from 0 to 999,999,999 inclusive.
+   */
+  constructor(seconds: Long, nanoseconds: Int) {
+    validateRange(seconds, nanoseconds)
+
+    this.seconds = seconds
+    this.nanoseconds = nanoseconds
+  }
+
+  constructor(source: Parcel) : this(source.readLong(), source.readInt())
+
+  @RequiresApi(Build.VERSION_CODES.O) constructor(time: Instant) : this(time.epochSecond, time.nano)
+
+  constructor(date: Date) {
+    val (seconds, nanoseconds) = date.toPreciseTime()
+
+    validateRange(seconds, nanoseconds)
+
+    this.seconds = seconds
+    this.nanoseconds = nanoseconds
+  }
+
+  /**
+   * Returns a new [Date] corresponding to this timestamp.
+   *
+   * This may lose precision.
+   */
+  fun toDate(): Date = Date(seconds * 1_000 + (nanoseconds / 1_000_000))
+
+  /** Returns a new [Instant] that matches the time defined by this timestamp. */
+  @RequiresApi(Build.VERSION_CODES.O)
+  fun toInstant(): Instant = Instant.ofEpochSecond(seconds, nanoseconds.toLong())
+
+  override fun compareTo(other: Timestamp): Int =
+    compareValuesBy(this, other, Timestamp::seconds, Timestamp::nanoseconds)
+
+  override fun equals(other: Any?): Boolean =
+    when {
+      (other === this) -> true
+      (other !is Timestamp) -> false
+      else -> compareTo(other) == 0
+    }
+
+  override fun hashCode(): Int = 31 * seconds.hashCode() + nanoseconds
+
+  override fun toString(): String = "Timestamp(seconds=$seconds, nanoseconds=$nanoseconds)"
+
+  override fun describeContents(): Int {
+    return 0
+  }
+
+  override fun writeToParcel(dest: Parcel, flags: Int) {
+    dest.writeLong(seconds)
+    dest.writeInt(nanoseconds)
+  }
+
+  companion object CREATOR : Parcelable.Creator<Timestamp> {
```

Comment:
Is it conventional to make the companion object the CREATOR? If not, it may make sense to instead use the companion object for a factory function to create a Timestamp from an Instant.

Reply:
It looks like the `CREATOR` is indeed conventional to be the companion object. LGTM.

Reply:
Well the creator needs to be implemented regardless, but we can add it as a field to the companion object (via `@JvmField`) like so:
```kotlin
companion object {
  @JvmField
  val CREATOR =
    object : Parcelable.Creator<Timestamp> {
      override fun createFromParcel(source: Parcel): Timestamp = Timestamp(source)
      override fun newArray(size: Int): Array<Timestamp?> = arrayOfNulls(size)
    }
}
```

Which I'm totally fine with, as either is fine. But I'm not sure I understand how that connects with providing a factory function for creating timestamps from instants? I might be misunderstanding something here.

Reply:
Looks like our comments got sent at the same time. I agree that having the CREATOR as the companion object is totally fine. No need to make any changes. I'm resolving this comment.